src/matrixmessageprocessor.ts

```diff
@@ -0,0 +1,281 @@
+import * as Discord from "discord.js";
+import { IMatrixMessage, IMatrixEvent } from "./matrixtypes";
+import * as Parser from "node-html-parser";
+import { Util } from "./util";
+import { DiscordBot } from "./bot";
+
+const MIN_NAME_LENGTH = 2;
+const MAX_NAME_LENGTH = 32;
+const MATRIX_TO_LINK = "https://matrix.to/#/";
+
+export class MatrixMessageProcessorOpts {
+    constructor(readonly disableEveryone: boolean = true, readonly disableHere: boolean = true) { }
+}
+
+export class MatrixMessageProcessor {
+    private guild: Discord.Guild;
+    private listDepth: number = 0;
+    private listBulletPoints: string[] = ["●", "○", "■", "‣"];
+    constructor(public bot: DiscordBot, public opts: MatrixMessageProcessorOpts) { }
+    public async FormatMessage(
+        msg: IMatrixMessage,
+        guild: Discord.Guild,
+        profile?: IMatrixEvent | null,
+    ): Promise<string> {
+        this.guild = guild;
+        this.listDepth = 0;
+        let reply = "";
+        if (msg.formatted_body) {
+            // parser needs everything wrapped in html elements
+            // so we wrap everything in <div> just to be sure stuff is wrapped
+            // as <div> will be un-touched anyways
+            const parsed = Parser.parse(`<div>${msg.formatted_body}</div>`, {
+                lowerCaseTagName: true,
+                pre: true,
+            // tslint:disable-next-line no-any
+            } as any);
+            reply = await this.walkNode(parsed);
+            reply = reply.replace(/\s*$/, ""); // trim off whitespace at end
+        } else {
+            reply = this.escapeDiscord(msg.body);
+        }
+
+        if (msg.msgtype === "m.emote") {
+            if (profile &&
+                profile.displayname &&
+                profile.displayname.length >= MIN_NAME_LENGTH &&
+                profile.displayname.length <= MAX_NAME_LENGTH) {
+                reply = `_${profile.displayname} ${reply}_`;
+            } else {
+                reply = `_${reply}_`;
+            }
+        }
+        return reply;
+    }
+
+    private escapeDiscord(msg: string): string {
+        if (this.opts.disableEveryone) {
+            msg = msg.replace(/@everyone/g, "@ everyone");
+        }
+        if (this.opts.disableHere) {
+            msg = msg.replace(/@here/g, "@ here");
+        }
+        msg = msg.replace(/@room/g, "@here");
```

Comment:
Question: This still potentially allows anyone to spam here on the discord side?

Reply:
yes, this way everyone who says `@room` would get that replaced to `@here`. Soru thought the proper permission parsing stuffs would fit better in a different PR

Reply:
it's in #328